### ISSUE TITLE
As a software developer of scientific software, I need a native Haskell implementation of the dcopy function so that I can copy elements from one vector into another in a type-safe, thread safe manner.

### ISSUE THREAD

**Issue by dlewissandy** [dlewissandy/lambda-blas#71]
Documentation for the dcopy function can be found at BLAS.

- [x] create native implementation of function
- [x] create haddock documentation of function
- [x] create unit test to demonstrate byte equivalence to FORTRAN BLAS library function
- [x] create benchmark(s) to quantify time performance relative to FORTRAN BLAS library function
- [x] demonstrate that native has a relative time performance of no more than 100x the FORTRAN blas function.

**Comment by dlewissandy** [comment]
Implementation Strategy:
* This function can be implemented easily by polymorphising the single precision version.

**Comment by dlewissandy** [comment]
Attached is the raw benchmark data:
[dcopy.zip](https://github.com/dlewissandy/lambda-blas/files/868043/dcopy.zip)

**Comment by dlewissandy** [comment]
# DCOPY BENCHMARK ANALYSIS #

## Special Case inc=1 ##
When the increment of traversal is equal to unity, the time performance of both the unsafe and pure implementations are well described by a linear model:

### PURE ###
```
lm(formula = PURE ~ N, data = dat[dat$INC == 1, ])

Residuals:
     Min       1Q   Median       3Q      Max
-0.63879 -0.06780 -0.04894 -0.02388  1.30170

Coefficients:
             Estimate Std. Error t value Pr(>|t|)
(Intercept) 1.381e-01  6.393e-02    2.16   0.0373 *
N           5.097e-03  9.717e-06  524.51   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.3607 on 37 degrees of freedom
Multiple R-squared:  0.9999,	Adjusted R-squared:  0.9999
F-statistic: 2.751e+05 on 1 and 37 DF,  p-value: < 2.2e-16
```
### UNSAFE ###
```
lm(formula = UNSAFE ~ N, data = dat[dat$INC == 1, ])

Residuals:
    Min      1Q  Median      3Q     Max
-1.1269 -0.1259  0.2040  0.2309  1.0076

Coefficients:
              Estimate Std. Error t value Pr(>|t|)
(Intercept) -1.678e-01  8.264e-02  -2.031   0.0495 *
N            1.114e-03  1.256e-05  88.710   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.4662 on 37 degrees of freedom
Multiple R-squared:  0.9953,	Adjusted R-squared:  0.9952
F-statistic:  7870 on 1 and 37 DF,  p-value: < 2.2e-16
```
### ASYMPTOTIC PERFORMANCE ###
In the limit as n becomes very large, the relative time performance should approach:
```
5.097e-03
-------- = 4.6x
1.114e-3
```
which is within the 100x performance threshold for this project milestone

## Special Case inc/=1 ##
When the increment of traversal is not equal to unity, the time performance of both the unsafe and pure implementations are well described by a multi-linear model:

### PURE ###
```
lm(formula = PURE ~ N + Z, data = dat[dat$INC != 1, ])

Residuals:
    Min      1Q  Median      3Q     Max
-0.8715 -0.2260  0.1232  0.2878  0.8229

Coefficients:
              Estimate Std. Error t value Pr(>|t|)
(Intercept) -1.676e-01  6.722e-02  -2.493   0.0162 *
N            1.435e-02  1.568e-04  91.508   <2e-16 ***
Z            7.656e-04  1.199e-05  63.878   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.4086 on 48 degrees of freedom
Multiple R-squared:  0.9989,	Adjusted R-squared:  0.9989
F-statistic: 2.183e+04 on 2 and 48 DF,  p-value: < 2.2e-16
```
### UNSAFE ###
```
lm(formula = UNSAFE ~ N + Z, data = dat[dat$INC != 1, ])

Residuals:
     Min       1Q   Median       3Q      Max
-0.82619 -0.19493  0.09707  0.22559  0.74187

Coefficients:
              Estimate Std. Error t value Pr(>|t|)
(Intercept) -1.986e-01  5.880e-02  -3.378  0.00145 **
N            4.920e-03  1.372e-04  35.867  < 2e-16 ***
Z            4.570e-04  1.048e-05  43.596  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.3574 on 48 degrees of freedom
Multiple R-squared:  0.9958,	Adjusted R-squared:  0.9956
F-statistic:  5689 on 2 and 48 DF,  p-value: < 2.2e-16
```
### ASYMPTOTIC PERFORMANCE ###
In the limit as both n and inc become very large, the relative time performance should approach:
```
7.656e-04
--------- = 1.7x
4.570e-04
```
which is within the 100x performance threshold for this project milestone.

**Comment by dlewissandy** [comment]
Closed as DONE after approval of pull request #80.
